rfcs/20230517-forward-compatibility-testing.md

```diff
@@ -0,0 +1,102 @@
+* **R3.** Test job must scale with releases and execute in a reasonable amount
+of time.
+
+## Preferred Design: Serialize at HEAD, Deserialize at Target Release
+
+_Note: Section only discusses how to test a generic release `0.X.0`. Discussion
+on what values of `X` to test for are discussed in the later section "What
+releases should be tested?"_
+
+### Serialize and Commit Test Files at HEAD
+
+Prior to tagging each StableHLO release, which occurs ~2x per week, generate
+versioned forward compatibility tests using the
+`stablehlo_legalize_to_vhlo.0_X_0.mlir` test file as follows:
+
+```bash
+stablehlo-translate --serialize --target=0.X.0
+  stablehlo_legalize_to_vhlo.0_X_0.mlir >
+    stablehlo_legalize_to_vhlo.0_X_0.mlir.bc.HEAD
+```
+
+The output will be a bytecode file that targets `0.X.0` and should be runnable
+at release `0.X.0`. Once these files are generated and checked in, we can add
+the following RUN line to the source `.mlir` file to test serialization,
+deserialization, and FileCheck comparisons:
+
+```bash
+// RUN: stablehlo-translate --deserialize %s.bc.HEAD |
+  stablehlo-translate --serialize --target=0.X.0 |
+    stablehlo-opt --mlir-print-op-generic | FileCheck %s
+```
+
+**Proposal 1:** Maintain bytecode at HEAD to use for forward compatibility
+testing.
+
+### Add Forward Compatibility CI
+
+A CI job which blocks pull request merges to main can be added to ensure
+forward compatibility of PRs. To test version `0.X.0` the CI job does the
+following:
+
+* Checkout `openxla/stablehlo` at tag `v0.X.0`.
+* Checkout file at HEAD `stablehlo_legalize_to_vhlo.0_X_0.mlir.bc.HEAD`
+* Append FileCheck line to `stablehlo_legalize_to_vhlo.0_X_0.mlir`
+* Perform the existing build-and-test CI workflow.
+
+This will rebuild LLVM at the proper revision, rebuild StableHLO at the tagged
+release, and bring in the forward compatibility test to the CI workspace. Any
+failures in this test file will indicate a forward incompatibility at HEAD. A
+prototype of this CI job can be seen [here](https://github.com/GleasonK/stablehlo/pull/33).
+
+**Proposal 2:** Add CI which builds at tagged release commits to check forward
+compatibility.
+
+## What releases should be tested?
+
+Testing every single release would be sound, but more expensive and time
+consuming than needed, and doesn't scale well with multiple releases a week
+(_R3_). Given that all changes to the StableHLO opset (_R1_), as well as
+changes to serialization machinery like targeting a newer version of the MLIR
+bytecode format (_R2_), require bumping the minor version, there is a low
+chance of causing forward incompatibilities between different patch versions of
+StableHLO (i.e. `0.10.1 → 0.10.2`). Ensuring compatibility between HEAD and all
+minor releases (`0.9.0`, `0.10.0`, etc.) should provide thorough enough
+coverage of the two mentioned sources of forward incompatibilities.
+
+Currently, this means that we only serialize a file at HEAD for all supported
```

Comment:
What does "supported" mean? E.g. v0.9.0 is already 2.5 months old, i.e. is well outside the current forward compatibility window. Is it still supported?

Reply:
I suppose supported is up to us to define, since our compatibility guarantees are a minimum.

Given that one use of forward compatibility is to provide users adequate time to upgrade, `v0.9.0` is only allowed to be removed as a supported forward compatible version once `v0.10.0` is a month old. 

Until we explicitly prevent it and clean up the code, the infrastructure allows targeting `v0.9.0`. We should consider a process for this cleanup, i.e. how aggressively do we want to cleanup IR downgrades?

Reply:
Created https://github.com/openxla/stablehlo/issues/1631